firedrake/assemble.py

```diff
@@ -165,79 +185,165 @@ def create_assembly_callable(expr, tensor=None, bcs=None, form_compiler_paramete
     """
     if tensor is None:
         raise ValueError("Have to provide tensor to write to")
-    if mat_type == "matfree":
-        return tensor.assemble
-    loops = _assemble(expr, tensor=tensor, bcs=solving._extract_bcs(bcs),
-                      form_compiler_parameters=form_compiler_parameters,
-                      mat_type=mat_type,
-                      sub_mat_type=sub_mat_type,
-                      diagonal=diagonal,
-                      assemble_now=False)
-
-    loops = tuple(loops)
-
-    def thunk():
-        for kernel in loops:
-            kernel()
-    return thunk
-
-
-def get_matrix(expr, mat_type, sub_mat_type, *, bcs=None,
-               options_prefix=None, tensor=None, appctx=None,
-               form_compiler_parameters=None):
+    return functools.partial(assemble, expr,
+                             tensor=tensor,
+                             bcs=bcs,
+                             form_compiler_parameters=form_compiler_parameters,
+                             mat_type=mat_type,
+                             sub_mat_type=sub_mat_type,
+                             diagonal=diagonal,
+                             assembly_type="residual")
+
+
+def _get_assembly_rank(expr, diagonal):
+    """Return the assembly rank.
+
+    :arg expr: The expression being assembled.
+    :arg diagonal: If assembling a matrix is it diagonal?
+
+    :returns: The appropriate :class:`_AssemblyRank`.
+    """
+    rank = len(expr.arguments())
+    if diagonal:
+        assert rank == 2
+        return _AssemblyRank.VECTOR
+    if rank == 0:
+        return _AssemblyRank.SCALAR
+    if rank == 1:
+        return _AssemblyRank.VECTOR
+    if rank == 2:
+        return _AssemblyRank.MATRIX
+    raise AssertionError
+
+
+def _assemble_scalar(expr, scalar, bcs, opts):
+    """Assemble a scalar.
```

Comment:
Would it be more accurate (and clear) to say `Assemble a scalar from a 0-form`. Would at least be consistent with the `ValueError`s below